test_bottles.py

```diff
@@ -1,90 +1,93 @@
-            '98 bottles of beer on the wall.\n'
-        )
-        self.assertEqual(Bottles().verse(99), expected)
-
-    def test_another_verse(self):
-        expected = (
-            '3 bottles of beer on the wall, '
-            '3 bottles of beer.\n'
-            'Take one down and pass it around, '
-            '2 bottles of beer on the wall.\n'
-        )
-        self.assertEqual(Bottles().verse(3), expected)
-
-    def test_verse_2(self):
-        expected = (
-            '2 bottles of beer on the wall, '
-            '2 bottles of beer.\n'
-            'Take one down and pass it around, '
-            '1 bottle of beer on the wall.\n'
-        )
-        self.assertEqual(Bottles().verse(2), expected)
-
-    def test_verse_1(self):
-        expected = (
-            '1 bottle of beer on the wall, '
-            '1 bottle of beer.\n'
-            'Take it down and pass it around, '
-            'no more bottles of beer on the wall.\n'
-        )
-        self.assertEqual(Bottles().verse(1), expected)
-
-    def test_verse_0(self):
-        expected = (
-            'No more bottles of beer on the wall, '
-            'no more bottles of beer.\n'
-            'Go to the store and buy some more, '
-            '99 bottles of beer on the wall.\n'
-        )
-        self.assertEqual(Bottles().verse(0), expected)
-
-    def test_a_couple_of_verses(self):
-        expected = (
-            '99 bottles of beer on the wall, '
-            '99 bottles of beer.\n'
-            'Take one down and pass it around, '
-            '98 bottles of beer on the wall.\n'
-            '\n'
-            '98 bottles of beer on the wall, '
-            '98 bottles of beer.\n'
-            'Take one down and pass it around, '
-            '97 bottles of beer on the wall.\n'
-        )
-        self.assertEqual(Bottles().verses(99, 98), expected)
-
-    def test_a_few_verses(self):
-        expected = (
-            '2 bottles of beer on the wall, '
-            '2 bottles of beer.\n'
-            'Take one down and pass it around, '
-            '1 bottle of beer on the wall.\n'
-            '\n'
-            '1 bottle of beer on the wall, '
-            '1 bottle of beer.\n'
-            'Take it down and pass it around, '
-            'no more bottles of beer on the wall.\n'
-            '\n'
-            'No more bottles of beer on the wall, '
-            'no more bottles of beer.\n'
-            'Go to the store and buy some more, '
-            '99 bottles of beer on the wall.\n'
-        )
-        self.assertEqual(Bottles().verses(2, 0), expected)
-
-    def test_the_whole_song(self):
-        expected = """99 bottles of beer on the wall, 99 bottles of beer.
+def test_the_first_verse():
+    expected = (
+        '99 bottles of beer on the wall, '
+        '99 bottles of beer.\n'
+        'Take one down and pass it around, '
+        '98 bottles of beer on the wall.\n'
+    )
+    assert Bottles().verse(99) == expected
```

Comment:
In my experience teaching Python beginners as well as working in our industry, Pytest should be chosen over unittest by default.  Outside of academia (not their job to code well) and folks coming from other languages (Java), I haven't seen unittest used.  My hunch is it's more of a case of those groups not being aware of pytest to start out with. :man_shrugging: 

Some high-level items to consider:

- most developers will know Pytest -> easier read for the market audience
- inexperienced readers will have an easier learning curve (eg no need for multiple assert methods)
- the modern pytest features may come in handy - marks, fixtures, parameterize
- despite being written in Python overall, the use of unittest would yet stymie my confidence in recommending this book or leveraging it in future teaching opportunities. :cry:

Reply:
Remember that this book is explicitly not for Python beginners. It's for Python developers who are new to OOP. We're explicitly *not* teaching Python.

> the modern pytest features may come in handy - marks, fixtures, parameterize

Do you mean that it will come in handy in the context of the book? Can you point out exactly where in the book Pytest will be better than unittest?

I'm not set against Pytest, I'm just not seeing the benefit here when the only difference is

```
assert Bottles().verse(99) == expected
```

vs 

```
self.assertEqual(Bottles().verse(3), expected)
```

Which to my eyes are saying _exactly_ the same thing, in two different ways that are both extremely common.

Reply:
I don't expect any of these to be used in the book's examples, but in case a brief overview is helpful here are the three features I mentioned.  While I don't foresee a use-case, maybe you will :blush: :man_shrugging: 

- fixtures: the streamlined way to provide setup for tests while ensuring clean teardowns when tests complete or fail
  - examples: create an infrastructure such as a database or filesystem, provide a Stub/Mock system to avoid I/O calls, or simply do a complex setup for a more hairy test (many required args perhaps)
- parametrize: de-duplicate tests. It's like meta-creating them on the fly, or defining them as a product
  - example: instead of writing a test for each possible value of arg, define the possible values and let it make a test for each pair of possibilities
- marks: allow you to group and run tests by a category, separate from the behaviour or module they are defined for.
  - examples: tests that are slow-running, require network access, take up large resources, etc

Reply:
Thanks. Yeah, I don't see any of this coming up in the context of 99 bottles :)